week2/game_2048/class_version/game_template.py

```diff
@@ -0,0 +1,230 @@
+        """
+        self._grid = grid
+
+    def set_tile(self, row, col, value):
+        """
+        Set the tile at position row, col to have the given value.
+        """
+        self._grid[row][col] = value
+
+    def get_tile(self, row, col):
+        """
+        Return the value of the tile at position row, col.
+        """
+        return self._grid[row][col]
+
+    def reset(self):
+        """
+        Reset the game so the grid is empty except for two
+        initial tiles.
+        """
+        #self.set_grid(self.generate_empty_grid())
+        [self.new_tile() for dummy_num in range(2)]
+
+    def generate_empty_grid(self):
+        """
+        Generates an empty grid of the size
+        _grid_height by _grid_width
+        """
+        return [[ 0 for dummy_col in range(self._grid_width)] for dummy_row in range(self._grid_height)]
+
+    def __str__(self):
+        """
+        Return a string representation of the grid for debugging.
+        """
+        return str(self.get_grid()).replace('],', '],\n')
+
+    def set_direction(self, value):
+        """
+        Sets the direction of the move
+        """
+        self._direction = value
+
+    def get_direction(self):
+        """
+        Returns the direction of the move
+        """
+        return self._direction
+
+    def move(self, direction):
+        """
+        Move all tiles in the given direction and add
+        a new tile if any tiles moved.
+        """
+        self.set_direction(direction)
+
+        if direction in (UP, DOWN):        
+            self.move_tiles(self.get_grid_height())
+        else:
+            self.move_tiles(self.get_grid_width())
+
+    def move_tiles(self, height_width):
+        """
+        Moves tiles to their new location on the grid
+        """
+        new_grid = self.generate_empty_grid()
+        for row in self.generate_index_grid(height_width):
+            self.add_merged_to_grid(row, new_grid)
+        #if self.get_grid() != new_grid:
+        #    self.set_grid(new_grid)
+        #    self.new_tile()
+        self.set_grid(new_grid)
+
+    def add_merged_to_grid(self, row, new_grid):
+        """
+        Calls merge on each row and adds it to a placeholder grid
+        """
+        temp_row = merge([self.get_tile(pair[0],pair[1]) for pair in row])
+        for index, value in enumerate(row):
+            new_grid[value[0]][value[1]] = temp_row [index]
+
+    def generate_index_grid(self, height_width):
+        """
+        Generates a grid of the indices
+        """
+        return [self.add_to_indices_grid(index, height_width) 
+                for index in self.get_move_grid_indices()[self.get_direction()]]
+
+    def add_to_indices_grid(self, index, height_width):
+        """
+        Adds each row of indices to the grid
+        """
+        return [index] + self.generate_row_indices(index[0], index[1], height_width)
+
+    def generate_row_indices(self, index0, index1, height_width):
```

Comment:
I want to be able to make this a comprehension... but I can't quite sort it out. I was thinking something like this:
```
    def generate_row_indices(self, index0, index1, height_width):
        return [self.get_indices(index0, index1) for dummy_number in range(height_width - 1)]

    def get_indices(self, index0, index1):
        index0 += OFFSETS[self.get_direction()][0]
        index1 += OFFSETS[self.get_direction()][1]
        return index0, index1

```
Should I continue working on it? Or does it seem okay as is?

Reply:
What you have is a correct refactoring of this. While above I said to prefer list comprehensions to for loops, for loops come in really handy when you want to generate lists in a memory-efficient way. What you could do here instead is to return a `generator` which is a function that runs up to the `yield` statement:
```python
def row_indices(self, index0, index1, height_width):
    for dummy_number in range(height_width - 1):
        index0 += OFFSETS[self.get_direction()][0]
        index1 += OFFSETS[self.get_direction()][1]
        yield index0, index1
```

To use this you would do:
```
for row_index in self.row_indices(index0, index1, height_width)
    # do something with row_index
```